Lib/ast.py

```diff
@@ -305,28 +306,17 @@ def get_docstring(node, clean=True):
     return text
 
 
-def _splitlines_no_ff(source):
+_line_pattern = re.compile(r"(.*?(?:\r\n|\n|\r|$))")
+def _splitlines_no_ff(source, maxlines=None):
     """Split a string into lines ignoring form feed and other chars.
 
     This mimics how the Python parser splits source code.
     """
-    idx = 0
     lines = []
-    next_line = ''
-    while idx < len(source):
-        c = source[idx]
-        next_line += c
-        idx += 1
-        # Keep \r\n together
-        if c == '\r' and idx < len(source) and source[idx] == '\n':
-            next_line += '\n'
-            idx += 1
-        if c in '\r\n':
-            lines.append(next_line)
-            next_line = ''
-
-    if next_line:
-        lines.append(next_line)
+    for lineno, match in enumerate(_line_pattern.finditer(source), 1):
+        if maxlines is not None and lineno > maxlines:
+            break
+        lines.append(match[0])
```

Comment:
Could you pull the None check out, eg:

```suggestion
    if maxlines is None:
        return [m[0] for m in _line_pattern.finditer(source)]
    for lineno, match in enumerate(_line_pattern.finditer(source), start=1):
        if lineno > maxlines:
            break
        lines.append(match[0])
```

I'm not sure if it's nicer though, it would be nicer to use `return pat.findall(s)` but we need to filter just to the first tuple element.

A

Reply:
I think the readability is actually worse after the change. We would have two very similar logics in two places. The potential benefit is performance - but it's not observable in the benchmark. The `is not None` is negligible compared to the regex checking.

Personally I think the current implementation is easier to understand.

Reply:
Agreed, the current is much easier to understand / reason about.